### ISSUE TITLE
Avoiding dependencies on config.h

### ISSUE THREAD

**Issue by jerryz123** [riscv/riscv-isa-sim#1168]
Currently many spike-installed headers depend on the autoconf-generated `config.h`, which prevents usage of spike-as-a-library when including those headers.

I would like to remove these dependencies so I can just include `sim.h` in an external project and link with the spike library.
What is the canonical way to do this?

**Comment by scottj97** [comment]
Linking Spike with outside code was enabled by a recent addition to this repo but is not well specified or well supported.

I link Spike with python but I use only a very small & simple C interface (that I added to my local fork of Spike) so I don't need any header files.

Spike compilation uses several configure-time choices such as `--enable-commitlog` and I assume this is why `config.h` is included. But perhaps the header files themselves don't need to care?

In the case of `sim.h` the interface changes if Boost libraries are available (`HAVE_BOOST_ASIO`). This has been a source of [issues](https://github.com/riscv-software-src/riscv-isa-sim/issues/834#issuecomment-944425960) in the past.

What if we moved all such `#ifdef` statements into the `*.cc` files and removed the include of `config.h` from the headers. Would that help you?
